### PR TITLE
Use Schema references, not Python references (#11)

### LINKED ISSUE #11: Use schema references, not Python references
Reused object definitions in I/O schema are linked on Python level. This makes the resulting schema JSON harder to read, as the reused definitions are simply copied there. Instead, define and reuse them on the schema level.

E.g. the [v4 schema](https://github.com/kernelci/kcidb-io/blob/main/kcidb_io/schema/v4.py) stores schema for each separate object type in Python variables `JSON_CHECKOUT`, `JSON_BUILD`, and `JSON_TEST`, and then references them when defining the complete schema in `JSON`. Similarly, `JSON_RESOURCE` is referenced in multiple places.

Instead, we should define and (re)use them within the schema itself, as described in [8.2.4. Schema Re-Use With "$defs"](https://json-schema.org/draft/2020-12/json-schema-core.html#defs) of the JSON Schema specification. We can still use the Python variables for storing the pieces, if needed, but linking and reusing should be done in the schema itself.

When changing this, take note of how `kcidb.orm` references the particular object type schemas and make sure they keep working after the change.

### PR DESCRIPTION
**Pull request by mrbazzan**
Closes #11

### CODE REVIEW

File: kcidb_io/schema/v1.py [hunk 1]
```diff
@@ -455,6 +455,410 @@
...
+            },
+            "additionalProperties": False,
+
+            "description":
+                "A revision of the tested code.\n"
+                "\n"
+                "Represents a way the tested source code could be obtained. E.g. "
+                "checking out a particular commit from a git repo, and applying a "
+                "set of patches on top.",
+            "$id": "/schemas/revision",
```
[the remark this answers is not in the document]
Comment (mrbazzan) [reply kcidb_io/schema/v1.py:617]: `$id` is used to define cannonical URI with pointer fragment. [check it out](https://json-schema.org/draft/2019-09/json-schema-core.html#idExamples)
Perusing [this document](https://json-schema.org/understanding-json-schema/structuring.html#id20) gave me the idea for using `$id`

[the remark this answers is not in the document]
Comment (mrbazzan) [reply kcidb_io/schema/v1.py:617]: Re-reading the doc shows that there was no mention of using `$id` in `$def`. I just inferred. I'll make the changes.

File: kcidb_io/schema/v1.py [hunk 2]
```diff
@@ -561,7 +965,7 @@ def get_version(data):
             except KeyError:
                 minor = 0
             if isinstance(major, int) and isinstance(minor, int) and \
-               major >= 0 and minor >= 0:
+                    major >= 0 and minor >= 0:
```
[the remark this answers is not in the document]
Comment (mrbazzan) [reply kcidb_io/schema/v1.py]: Pycharm reformat is responsible for this. I'll revert.

[the remark this answers is not in the document]
Comment (spbnick) [reply kcidb_io/schema/v1.py]: It might be a good idea to do these, and to add a step to check the indents are good to the CI. However, that needs to be done in a separate commit, or a separate PR. Thanks!

File: kcidb_io/schema/v1.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (spbnick) [reply kcidb_io/schema/v1.py:205]: Good job figuring this all out, and I understand this could be useful if we want to keep schemas in separate files and then bundle them automatically. However, we do **not** want to put the schema in separate files so far (i.e. we don't want to "unbundle" it). It's easier to deal with it in a single file thus far.

Using this mechanism would be clever, but too involved for our needs right now. It is not obvious how this works, and reading and understanding the article you linked takes effort and time. So let's not subject others and future us to this unnecessarily, and use the relatively-obvious fragment URIs pointing to the schemas using "JSON Pointers". I.e. just put schemas under `"$defs"`, and use `"$ref": "#/$defs/resource"` and similar.

We control the entirety of the schema, and we would be able to change the references and IDs if we ever need it.
Thank you.

[the remark this answers is not in the document]
Comment (mrbazzan) [reply kcidb_io/schema/v1.py:205]: The article does take a lot of effort and time. I'll submit a commit to that effect.

Also, what should I do about the failing test?

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,4 @@
+.idea/
```
Comment (spbnick) [inline .gitignore:1]: If this change is intentional, could you please put it into a separate commit?

File: kcidb_io/schema/v1.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (spbnick) [reply kcidb_io/schema/v1.py:205]: > Also, what should I do about the failing test?

Wrap the lines. Rewrap the documentation paragraphs as necessary. Are there any particular lines that are giving you trouble?

If this turns out too difficult or ugly, we could still just leave the schema pieces in the variables as they are, and paste *them* into `$defs` instead.

File: .gitignore (hunk 3 above)
Comment (mrbazzan) [reply .gitignore:1]: Done!

File: kcidb_io/schema/v4.py [hunk 4]
```diff
@@ -29,582 +29,606 @@
...
-                "The URL of a plain-text log file of the checkout attempt. "
-                "E.g. 'git am' output.",
-        },
-        "log_excerpt": {
-            "type": "string",
-            "maxLength": 16384,
-            "description":
-                "A part of the log file of the checkout attempt most "
-                "relevant to its outcome.",
+            "description": "A named remote resource",
```
Comment (spbnick) [inline kcidb_io/schema/v4.py:89]: Could you keep the `"description"` property where it was, right below the `"title"`, for each object?

File: kcidb_io/schema/v4.py [hunk 5]
```diff
@@ -29,582 +29,606 @@
...
+                        "./arch/arm64/include/asm/pointer_auth.h:55:2: "
+                        "note: expanded from macro '__ptrauth_key_install'\n"
+                        "        write_sysreg_s(__pki_v.lo, SYS_ ## k "
+                        "## KEYLO_EL1);     \\\n"
+                        "        ^\n"
+                        "./arch/arm64/include/asm/sysreg.h:829:37: note: "
+                        "expanded from macro 'write_sysreg_s'\n asm "
+                        "volatile(__msr_s(r, \"%x0\") : : \"rZ\" "
+                        "(__val));               \\\n"
+                        "                           ^\n"
```
Comment (spbnick) [inline kcidb_io/schema/v4.py:429]: This example lost some of its formatting. It should be this instead:
```suggestion
                        "In file included from "
                        "./arch/arm64/include/asm/processor.h:35:\n"
                        "./arch/arm64/include/asm/pointer_auth.h:62:3: "
                        "error: invalid input constraint 'rZ' in asm\n"
                        "                __ptrauth_key_install(APIA, "
                        "keys->apia);\n                ^\n"
                        "./arch/arm64/include/asm/pointer_auth.h:55:2: "
                        "note: expanded from macro '__ptrauth_key_install'\n"
                        "        write_sysreg_s(__pki_v.lo, "
                        "SYS_ ## k ## KEYLO_EL1);     \\\n"
                        "        ^\n./arch/arm64/include/asm/sysreg.h:829:37:"
                        " note: expanded from macro 'write_sysreg_s'\n"
                        "        asm volatile(__msr_s(r, \"%x0\") : : \"rZ\" "
                        "(__val));               \\\n"
                        "                                           ^\n"
```

**Review by spbnick — CHANGES_REQUESTED** [review]
We're almost there, thank you @mrbazzan!

I have a few inline requests, but also please squash your history into two commits: one adding the `.idea/` to `.gitignore`, and another doing the rest, both with appropriate commit subjects and messages.

Thank you!

**Comment by spbnick** [comment]
Looks and works great! Merging! Thank you, Abdulwasiu :smile:
